### PR TITLE
[Modes] fix handle_torch_funcion logic

### LINKED ISSUE #85696: TorchFunctionMode inside backward call is ignored when outer mode is present
### 🐛 Describe the bug

Changes introduced in https://github.com/pytorch/pytorch/pull/84774 broke the composition of function modes when used with `torch.autograd.Function`. Here's a reproducer:

```python
import torch

class ErrorMode(torch.overrides.TorchFunctionMode):
    def __torch_function__(
        self,
        orig_func,
        types,
        args = (),
        kwargs = None,
    ):
        if kwargs is None:
            kwargs = {}
        raise RuntimeError("test")
        return orig_func(*args, **kwargs)

class PassThrough(torch.overrides.TorchFunctionMode):
    def __torch_function__(
        self,
        orig_func,
        types,
        args = (),
        kwargs = None,
    ):
        if kwargs is None:
            kwargs = {}
        return orig_func(*args, **kwargs)

# Taken from
# https://pytorch.org/tutorials/beginner/examples_autograd/polynomial_custom_function.html
class LegendrePolynomial3(torch.autograd.Function):
    @staticmethod
    def forward(ctx, input):
        ctx.save_for_backward(input)
        return 0.5 * (5 * input ** 3 - 3 * input)

    @staticmethod
    def backward(ctx, grad_output):
        input, = ctx.saved_tensors
        with ErrorMode():
            result = grad_output * 1.5 * (5 * input ** 2 - 1)
        return result

P3 = LegendrePolynomial3.apply
x = torch.randn(2, requires_grad=True)

def func(x):
    y = P3(x)
    return torch.autograd.grad([y], [x], [torch.ones_like(y)])

try:
    func(x)
except RuntimeError as e:
    print(f"Error message: {e}")

# Should raise "RuntimeError: test" but doesn't
try:
    with PassThrough():
        func(x)
        print("No error!")
except RuntimeError as e:
    print(f"Error message: {e}")

```

Reverting https://github.com/pytorch/pytorch/pull/84774 fixes the problem.

### Versions

Latest master.

cc @hameerabbasi @rgommers @peterbell10 @ezyang @Chillee @zou3519 @albanD @samdow

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85707

Fixes #85696. I didn't totally get what was happening in handle_torch_function and so was trying to recreate the original logic instead of follow what the C++ is doing. This fixes that

### CODE REVIEW

File: torch/overrides.py [hunk 1]
```diff
@@ -1511,10 +1511,10 @@ def handle_torch_function(
...
     # Check for __torch_function__ mode.
     if _is_torch_function_mode_enabled():
-        mode = _get_current_function_mode()
-        # NB: unlike on tensors, modes are instances
-        with _no_torch_function_mode():
-            result = mode.__torch_function__(public_api, types, args, kwargs)
+        # if we're here, the mode must be set to a TorchFunctionStackMode
+        # this unsets it and calls directly into TorchFunctionStackMode's torch function
+        _set_torch_function_mode(None)
+        result = _TorchFunctionStackMode().__torch_function__(public_api, types, args, kwargs)
```
Comment (ezyang) [inline torch/overrides.py:1517]: The code smell here, is that the unsetting of the torch function mode is not scoped, so you're destroying whatever the previous state was by the time you come out. Maybe that's what the old code is, but it's not obvious to me that this is right either.

**Review by ezyang — APPROVED** [review]
I checked the implementation of TorchFunctionStackMode and I see it will reset torch function mode appropriately on exit via inner context manager, so I will approve the bugfix. However, it would be nice if locally it was obvious that dynamic scoping was respected (and it isn't possible for an incorrect direct set to leak out.)

**Comment by samdow** [comment]
Just added a try/finally around the call. As you found, I wasn't able to get a test that failed before and works after because the stack mode's context manager was taking care of state but agree that the local obivous-ness is nice

**Comment by samdow** [comment]
> Thanks for the modification here. I notice there are still other spots in the code that are using the bare setter pattern, so maybe in a separate PR it would be good to go over those sites as well.

Do you mean in this file or elsewhere? I think everywhere in this file is actually pretty safe but I can at minimum comment through why this is (and it actually shouldn't even be necessary if we're able to move this all to C++)

> I am actually a little perplexed why you can't use _no_torch_function_mode

Oh I see--yeah we totally could and that might make it easier to remember to use this pattern. I'll change it just so we don't have that duplicate code running around

> and it also seems weird that we have to construct a _TorchFunctionStackMode object to call its method lol

I'm trying to avoid exposing the _get_torch_function_mode function mostly as a safety thing as I was writing this (many places previously called _get_torch_{function|dispatch}_mode and now they're going to want the mode in the stack instead of the C++ mode). Since we know that the C++ mode is either going to be _TorchFunctionStackMode or nothing, it just recreates the stack modes sometimes instead of only ever having one C++ mode and passing it around (it never has state so this might be wasteful but is safe)

**Comment by ezyang** [comment]
Yeah, I'm referring to other files, mostly in _TorchFunctionStackMode. It looked for the most part that there weren't any real bugs, since eventually you hit it with a context manager or something.

**Pull request closed without merging**